### PR TITLE
feat(mdns): adaptive initial interval peer discovery

### LINKED ISSUE #3319: Manage an initial delay before the first mdns query is sent on an interface
## Description

Based on identify behaviour, manage an initial delay before the first mdns query is sent on an interface.

## Motivation

In a docker environment, the first mdns query can be lost if the interface is not ready yet. As the interval query is by default rather large (5 minutes), using an initial delay before the first mdns query is sent would improve the speed of peers discovery.

## Current Implementation

No way to configure an initial delay.

## Are you planning to do it yourself in a pull request?

Yes

### PR DESCRIPTION
**Pull request by stormshield-pj50**
## Description

Peer discovery with mDNS can be very slow, particularly if the first mDNS query is lost. This patch resolves it by adjusting the timer with an adaptive initial interval. We start with a very short timer (500 ms). If a peer is discovered before the end of the timer, then the timer is reset to the normal query interval value (300s), otherwise the timer's value is multiplied by 2 until it reaches the normal query interval value.

Related: https://github.com/libp2p/rust-libp2p/pull/3323.
Resolves: https://github.com/libp2p/rust-libp2p/issues/3319.

## Change checklist

- [X] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/mdns/src/behaviour/iface.rs [hunk 1]
```diff
@@ -217,6 +289,12 @@ where
 
                     self.discovered
                         .extend(response.extract_discovered(Instant::now(), self.local_peer_id));
+
+                    // Stop probing when we have a valid response
+                    if !self.discovered.is_empty() {
+                        self.probe_state.update(true, &self.query_interval);
```
Comment (thomaseizinger) [inline protocols/mdns/src/behaviour/iface.rs]: Wouldn't it be easier to understand to assign `probe_state` directly to a new variant here?

File: protocols/mdns/src/behaviour/iface.rs [hunk 2]
```diff
@@ -159,6 +222,16 @@ where
...
+
+                let is_probing = self.probe_state.is_probing();
+                self.probe_state
+                    .update(self.discovered.is_empty(), &self.query_interval);
+
+                // Reset if the previous state was Probing or if the state has been switched to Finished
+                // We do not need to reset the timer after a tick in Finished state, the reset should occur reset when the interface receive a response from a peer.
+                if is_probing {
+                    self.reset_timer();
+                }
```
Comment (thomaseizinger) [inline protocols/mdns/src/behaviour/iface.rs]: I think similarly, I'd find it easier to understand to do something like:

```rust
let continue_probing = self.discovered.is_empty();

if continue_probing {
	self.probe_state.increase_interval(); // Note: This is a no-op for `Finished` / once we hit `MAX_INTERVAL`
} else {
	self.probe_state = ProbeState::Finished(MAX_INTERVAL);
}
```

**Review by thomaseizinger — COMMENTED** [review]
Thank you!

Two comments. Let me know what you think.

File: protocols/mdns/src/behaviour/iface.rs (hunk 1 above)
Comment (stormshield-pj50) [reply protocols/mdns/src/behaviour/iface.rs]: I agree. Code is much simpler now

File: protocols/mdns/src/behaviour/iface.rs (hunk 2 above)
Comment (stormshield-pj50) [reply protocols/mdns/src/behaviour/iface.rs]: OK but I prefer inlining increase_interval() code, its easier to read

File: protocols/mdns/src/behaviour/iface.rs [hunk 3]
```diff
@@ -159,6 +186,25 @@ where
             if Pin::new(&mut self.timeout).poll_next(cx).is_ready() {
                 log::trace!("sending query on iface {}", self.addr);
                 self.send_buffer.push_back(build_query());
+                log::trace!("tick on {:#?} {:#?}", self.addr, self.probe_state);
+
+                let continue_probing = self.discovered.is_empty();
```
Comment (thomaseizinger) [inline protocols/mdns/src/behaviour/iface.rs]: Do we need this here if we set it on line 268 already?

I think it is easier to understand if we pair the changes to `probe_state` with the changes to `discovered`.

**Review by thomaseizinger — COMMENTED** [review]
Thanks! One more question

File: protocols/mdns/src/behaviour/iface.rs (hunk 3 above)
Comment (stormshield-pj50) [reply protocols/mdns/src/behaviour/iface.rs]: You're right the check on self.discovered when the timer triggers is useless

**Review by thomaseizinger — DISMISSED** [review]
This looks good to me, thank you!

I wonder if we should have an automated test for this? It is a bit tricky. If you have an idea, it would be much appreciated.

Can you confirm that the current version works as expected?

**Comment by thomaseizinger** [comment]
Clippy doesn't seem to be happy though.

**Comment by stormshield-pj50** [comment]
> This looks good to me, thank you!
> 
> I wonder if we should have an automated test for this? It is a bit tricky. If you have an idea, it would be much appreciated.
> 
> Can you confirm that the current version works as expected?

On our side we have an automated docker testsuite and the first mDNS query can be lost if the interface is not ready yet.

Anyway the current version works, we just did plug our manual libp2p test tool with it and the mdns peer discovery works  as expected,  both with a successful and an unsuccessful discovery, the probe ticks logs are OK.

**Comment by thomaseizinger** [comment]
> > This looks good to me, thank you!
> > I wonder if we should have an automated test for this? It is a bit tricky. If you have an idea, it would be much appreciated.
> > Can you confirm that the current version works as expected?
> 
> On our side we have an automated docker testsuite and the first mDNS query can be lost if the interface is not ready yet.
> 
> Anyway the current version works, we just did plug our manual libp2p test tool with it and the mdns peer discovery works as expected, both with a successful and an unsuccessful discovery, the probe ticks logs are OK.

The code looks good to me too. I am fine merging without an automated test. Thanks for confirming!

File: protocols/mdns/CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,7 @@
 - Raise MSRV to 1.65.
   See [PR 3715].
 - Remove deprecated `Mdns` prefixed items. See [PR 3699].
+- Faster peer discovery with adaptive initial interval. See [PR 3975].
```
Comment (mxinden) [inline protocols/mdns/CHANGELOG.md:10]:
```suggestion

[PR 3975]: https://github.com/libp2p/rust-libp2p/pull/3975
```

**Review by mxinden — DISMISSED** [review]

**Review by mxinden — APPROVED** [review]

**Pull request merged**